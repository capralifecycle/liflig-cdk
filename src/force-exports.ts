import * as cdk from "@aws-cdk/core"

function getStageOrApp(scope: cdk.Construct): cdk.Construct {
  const stage = cdk.Stage.of(scope)
  if (stage != null) {
    return stage
  }

  const app = scope.node.root
  if (!cdk.App.isApp(app)) {
    throw new Error(`Could not locate cdk App for ${scope.node.id}`)
  }
  return app
}

/**
 * The intention of this method is to force exports from the stack
 * without the need for another stack to depend on it.
 *
 * This helps us to solve two problems:
 *
 *   1. When we remove a resource that depends on the stack, it will
 *      first attempt to remove the export from the stack, causing
 *      the deployment to fail. Now it will leave the export.
 *      See also https://github.com/aws/aws-cdk/issues/7602
 *
 *   2. During development, we can deploy certain resources directly from
 *      localhost, without adding new exports to the stack which is
 *      managed by the CD pipeline.
 *
 * The stack produced is never supposed to be deployed. It should only
 * be used as hints for CDK.
 */
export class ForceExports extends cdk.Stack {
  private i = 0

  constructor(scope: cdk.Construct) {
    super(getStageOrApp(scope), `EXPORTS-${scope.node.id}`, {
      env: {
        account: cdk.Stack.of(scope).account,
        region: cdk.Stack.of(scope).region,
      },
    })
  }

  private genId() {
    return `Output${this.i++}`
  }

  public register(resource: string): void {
    new cdk.CfnOutput(this, this.genId(), {
      value: resource,
    })
  }
}
