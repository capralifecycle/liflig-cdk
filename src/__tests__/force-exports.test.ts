import { Vpc } from "@aws-cdk/aws-ec2"
import { App, Stack } from "@aws-cdk/core"
import "jest-cdk-snapshot"
import { ForceExports } from "../force-exports"

test("force-exports", () => {
  const app = new App()
  const stack1 = new Stack(app, "Stack1")

  const vpc = new Vpc(stack1, "Vpc")

  const forceExports = new ForceExports(stack1)
  forceExports.register(vpc.vpcId)
  for (const subnet of vpc.privateSubnets.concat(vpc.publicSubnets)) {
    forceExports.register(subnet.subnetId)
  }

  expect(forceExports).toMatchCdkSnapshot()
  expect(forceExports.node.id).toBe("EXPORTS-Stack1")
})
