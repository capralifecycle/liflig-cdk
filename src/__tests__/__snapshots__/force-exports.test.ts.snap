// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`force-exports 1`] = `
Object {
  "Outputs": Object {
    "Output0": Object {
      "Value": Object {
        "Fn::ImportValue": "Stack1:ExportsOutputRefVpc8378EB38272D6E3A",
      },
    },
    "Output1": Object {
      "Value": Object {
        "Fn::ImportValue": "Stack1:ExportsOutputRefVpcPrivateSubnet1Subnet536B997AFD4CC940",
      },
    },
    "Output2": Object {
      "Value": Object {
        "Fn::ImportValue": "Stack1:ExportsOutputRefVpcPrivateSubnet2Subnet3788AAA1380949A3",
      },
    },
    "Output3": Object {
      "Value": Object {
        "Fn::ImportValue": "Stack1:ExportsOutputRefVpcPublicSubnet1Subnet5C2D37C4FFA2B456",
      },
    },
    "Output4": Object {
      "Value": Object {
        "Fn::ImportValue": "Stack1:ExportsOutputRefVpcPublicSubnet2Subnet691E08A351552740",
      },
    },
  },
}
`;
